### DISCUSSION TITLE
AVR32DA32, millis()/micros() - problem

### DISCUSSION THREAD

**Discussion by hikssss** [SpenceKonde/DxCore#556] [Q&A]
Hi all, I made a board on AVR32DA32, installed kernel v.1.5.11, in Arduino IDE 2.3.3.
When using millis()/micros() functions - they do not work. Changed different settings and selected different timers from the menu - no help. What can you think of?

**Comment by grandaspanna** [comment]
Any chance you're not using setup()/loop() and instead using main()?

For elimination of any doubt, post a code example.

**Comment by hikssss** [comment]
![A](https://github.com/user-attachments/assets/d15ee361-1077-45fa-9446-91cfc0bafae6)

![b](https://github.com/user-attachments/assets/6349bfe2-fa06-45a0-8a5b-184e22207593)

**Reply by grandaspanna** [reply]
Looks reasonable - just wanted to eliminate one possibility. I don't see an obvious issue, but Arduino 2.X is not officially supported. Are you able to repeat the behaviour with 1.8.x? I've used some DA chips successfully...

**Comment by hikssss** [comment]
I'll try it on the 1.8 and post it here.
I also tried attiny826,- millis/micros/delay seem to work fine.

**Comment by hikssss** [comment]
nothing has changed (((
![c](https://github.com/user-attachments/assets/3a741e3b-ca80-487d-b167-87852c3a76af)

**Comment by grandaspanna** [comment]
That's pretty annoying. The 826 uses a different core. Which version of the Dx core do you have installed?

**Comment by hikssss** [comment]
V.1.5.11
![w](https://github.com/user-attachments/assets/e8c8fc9d-7f3d-428b-a745-8c74d5a410fd)

**Comment by grandaspanna** [comment]
Mmmm, ok. I've got a DA28 board that runs fine and uses millis() quite a bit. However, I'm running it at 16MHz (neopixel library doesn't seem to be happy with 24MHz). Also, I'm not using optiboot.

**Comment by hikssss** [comment]
OK, I'll try it now at 16mhz and no bootloader.

**Comment by hikssss** [comment]
New problem, when I program directly without bootloader, Serial.print does not work anymore....
Outputs the first letter of the message with high speed. I tried different variants and it didn't help. 

![a](https://github.com/user-attachments/assets/4f01ebea-d435-444c-a819-478f0a032e85)
![b](https://github.com/user-attachments/assets/1c35e32a-c095-4e3f-bf07-c759f419f206)

**Comment by hikssss** [comment]
I also have AVR128DB28 available, but it needs to be soldered on the board. I will do it a little later and check its behavior.

**Comment by hikssss** [comment]
I tested AVR128DB28 - everything is fine, both with bootloader and without bootloader everything works! I tried another AVR32DA32 - does not work normally. I think the problem is in DxCore, right?
![5](https://github.com/user-attachments/assets/fd0b7ff8-c0ed-4b88-9390-67fef2350610)
![6](https://github.com/user-attachments/assets/a25177ef-c39e-4a48-910c-8c4f19b846ed)

**Comment by fsievers** [comment]
Is it possible that you MCU is rebooting after one iteration of the loop?
Can you add a counter variable that increases on every loop() call and also print it's value?

**Comment by hikssss** [comment]
with a bootloader:
![1](https://github.com/user-attachments/assets/ad44597c-108e-4a9d-aa82-012c01967886)

without a bootloader:
![2](https://github.com/user-attachments/assets/2435b989-921b-4bb3-bbd5-8445a459c2fc)

**Comment by hikssss** [comment]
When loading a sketch without a bootloader, the MCU keeps resetting. This can be seen by a slightly glowing LED

```
void setup() {
  Serial.begin(9600);
  pinMode(LED_BUILTIN, OUTPUT);
  digitalWrite(LED_BUILTIN, HIGH);
}

void loop() {
  digitalWrite(LED_BUILTIN, LOW);
  static uint16_t a = 0;
  Serial.print("a: "); Serial.print(a++); Serial.println();
  Serial.print("millis: "); Serial.print(millis()); Serial.println();
  Serial.print("micros: "); Serial.print(micros()); Serial.println();
}
```

**Comment by hikssss** [comment]
Who else would check avr32da32, that would be great. But 99.9% it's a kernel bug.

**Comment by MX682X** [comment]
for me this looks like interrupts are not working correctly. Try adding this to setup:
```
  pinMode(PIN_PA6, OUTPUT);
  pinConfigure(PIN_PA7, (PIN_DIR_INPUT | PIN_PULLUP | PIN_ISC_FALL));
```

and that at the end of the sketch.
```
ISR(PORTA_PORT_vect) {
  digitalWrite(PIN_PA6, HIGH);
  VPORTA.INTFLAGS |= 0x80;
}
```

Substitute PORT letter and Pin Number to whatever is available. A button or Wire on Input and an LED or something on the output.

If it is indeed not working, try adding a 
sei();
to setup and check if it still isn't working. If it starts working, we'll know that the "I" flag in SREG is missing for some reason
